File: algorithms/sorting/mergesort.hs
mergesort :: Ord a => [a] -> [a]

mergesort = headOrEmpty . msort . map (\x -> [x])
    where
        -- merge two sorted lists
        merge :: Ord a => [a] -> [a] -> [a]
        merge [] ys = ys
        merge xs [] = xs
        merge (x:xs) (y:ys)
                | x < y     = x:merge xs (y:ys)
                | otherwise = y:merge (x:xs) ys

        headOrEmpty :: [[a]] -> [a]
        headOrEmpty [x] = x
        headOrEmpty [] = []

        -- merge sort a list of singletons using merge
        msort :: Ord a => [[a]] -> [[a]]
        msort [] = []
        msort [x] = [x]
        msort (x:x':xs) = msort ((merge x x'):xs)

File: algorithms/sorting/mergesort.js
'use strict';

const mergesort = (function () {
    let sorted;

    function merge(from, middle, to, array) {
        let i = from,
            j = middle;

        while (i < middle && j < to) {
            if (array[i] < array[j]) {
                sorted[from] = array[i];
                i += 1;
            } else {
                sorted[from] = array[j];
                j += 1;
            }

            from += 1;
        }

        while (i < middle) {
            sorted[from] = array[i];
            i += 1;
            from += 1;
        }

        while (j < to) {
            sorted[from] = array[j];
            j += 1;
            from += 1;
        }
    }

    function mergesort(from, to, array) {

        if (from >= (to - 1)) {
            return;
        }

        const middle = ((to - from) >> 1) + from;

        mergesort(from, middle, array);
        mergesort(middle, to, array);

        merge(from, middle, to, array);

        while (from < to) {
            array[from] = sorted[from];
            from += 1;
        }
    }

    return function (array) {
        sorted = [];

        mergesort(0, array.length, array);

        return array;
    }
} ());

const test = [0, 1, 100, -3, 2, 6, -2, -1, 0, 33, -7, 68, 19];
mergesort(test);

console.log(test);

const test2 = [0, 1, 100, -37, 27, 67, -27, -21, 10, 323, -7, 68, 19];
mergesort(test2);

console.log(test2);